apiserver/tools.go

```diff
@@ -102,100 +128,104 @@ func (h *toolsUploadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-		// them and cache in tools storage.
+		// so look for them in simplestreams,
+		// fetch them and cache in tools storage.
 		logger.Infof("%v agent binaries not found locally, fetching", version)
-		reader, err = h.fetchAndCacheTools(version, storage, st)
+		md, reader, err = h.fetchAndCacheTools(version, storage, st)
 		if err != nil {
 			err = errors.Annotate(err, "error fetching agent binaries")
 		}
 	}
 	if err != nil {
-		return nil, err
-	}
-	defer reader.Close()
-	data, err := ioutil.ReadAll(reader)
-	if err != nil {
-		return nil, errors.Annotate(err, "failed to read agent binaries tarball")
+		storage.Close()
+		return nil, 0, err
 	}
-	return data, nil
+
+	return &toolsReadCloser{f: reader, st: storage}, md.Size, nil
 }
 
 // fetchAndCacheTools fetches tools with the specified version by searching for a URL
 // in simplestreams and GETting it, caching the result in tools storage before returning
 // to the caller.
-func (h *toolsDownloadHandler) fetchAndCacheTools(v version.Binary, stor binarystorage.Storage, st *state.State) (io.ReadCloser, error) {
+func (h *toolsDownloadHandler) fetchAndCacheTools(
+	v version.Binary,
+	stor binarystorage.Storage,
+	st *state.State,
+) (binarystorage.Metadata, io.ReadCloser, error) {
+	md := binarystorage.Metadata{Version: v.String()}
+
 	newEnviron := stateenvirons.GetNewEnvironFunc(environs.New)
 	env, err := newEnviron(st)
 	if err != nil {
-		return nil, err
+		return md, nil, err
 	}
 	tools, err := envtools.FindExactTools(env, v.Number, v.Series, v.Arch)
 	if err != nil {
-		return nil, err
+		return md, nil, err
 	}
 
 	// No need to verify the server's identity because we verify the SHA-256 hash.
 	logger.Infof("fetching %v agent binaries from %v", v, tools.URL)
 	resp, err := utils.GetNonValidatingHTTPClient().Get(tools.URL)
 	if err != nil {
-		return nil, err
+		return md, nil, err
 	}
 	defer resp.Body.Close()
 	if resp.StatusCode != http.StatusOK {
 		msg := fmt.Sprintf("bad HTTP response: %v", resp.Status)
 		if body, err := ioutil.ReadAll(resp.Body); err == nil {
 			msg += fmt.Sprintf(" (%s)", bytes.TrimSpace(body))
 		}
-		return nil, errors.New(msg)
+		return md, nil, errors.New(msg)
 	}
+
 	data, sha256, err := readAndHash(resp.Body)
 	if err != nil {
-		return nil, err
+		return md, nil, err
 	}
 	if int64(len(data)) != tools.Size {
-		return nil, errors.Errorf("size mismatch for %s", tools.URL)
+		return md, nil, errors.Errorf("size mismatch for %s", tools.URL)
 	}
+	md.Size = tools.Size
 	if sha256 != tools.SHA256 {
-		return nil, errors.Errorf("hash mismatch for %s", tools.URL)
+		return md, nil, errors.Errorf("hash mismatch for %s", tools.URL)
 	}
+	md.SHA256 = tools.SHA256
 
-	// Cache tarball in tools storage before returning.
-	metadata := binarystorage.Metadata{
-		Version: v.String(),
-		Size:    tools.Size,
-		SHA256:  tools.SHA256,
+	if err := stor.Add(bytes.NewReader(data), md); err != nil {
+		return md, nil, errors.Annotate(err, "error caching agent binaries")
```

Comment:
So in this condition, we *do* buffer the full content first, right?
Its still better in the general case, so probably fine.

Reply:
Yeah, it would be a bit trickier here. 
We want the size and the hash for storage with the metadata.
Some chicanery io.Multiwriter might get us there, but it is looks like lukewarm (tops) RoI.